src/web/src/App.vue

```diff
@@ -13,17 +13,30 @@ export default {
   name: "App",
   components: {},
   async created() {
+    const darkModeQueryBody = "(prefers-color-scheme: dark)";
+    if (this.$cookies.get("darkMode")  === "true" ||
+        (this.$cookies.get("darkMode") ===  null  &&
+          window.matchMedia(darkModeQueryBody).matches)) {
+      this.$store.commit(TOGGLE_DARK_MODE);
+    }
+
+    if (window.matchMedia &&
+        window.matchMedia(darkModeQueryBody).media !== 'not all') {
+      const queryColorScheme = window.matchMedia(darkModeQueryBody);
+      queryColorScheme.addEventListener('change', () => {
+        if (this.$cookies.get("darkMode") === null) {
+          this.$store.commit(TOGGLE_DARK_MODE);
+        }
+      });
+    }
```

Comment:
From what I understand, it looks like the first `if` block is checking the cookies to see if the user has an existing color-scheme preference and the second `if` block is setting the colorscheme to the device preference. right?

If so, lets maybe refactor a bit. Here's a few ideas so that we can get to:
```js
async created() {
  this.syncDarkMode();
  ...
}
syncDarkMode() {
  if (this.isDarkModeCookieSet()) {
     this.syncDarkModeWithCookie();
  } else {
     this.syncDarkModeWithDevicePreferance();
  }
}
```
this way the code reads a bit simpler and we can seperate the concerns of checking if the dark mode is set, linking to cookie and matching device preference. 

wyt?

Reply:
Thanks for your comments!
Yes, I think your refactoring ideas are great.  Semantic function names are intuitive.  

There are still two points:
1. The first `if` checks the cookie or current scheme, and the second `if` checks the native support of color scheme: both `if` blocks will run, so it cannot be considered as a `if-else` block.
2. [Most browsers](https://caniuse.com/mdn-api_window_matchmedia) support `window.matchMedia`, so I deleted the testing for capabilities.

How about this:  `App.Vue`
```javascript
  async created() {
    this.darkModeInit();

    const querySemester = this.$route.query.semester;
    this.selectedSemester =
    // ......
  },
  methods: {
    darkModeInit() {
      this.syncColorScheme();
      this.registerColorSchemeListener();
    },
    syncColorScheme() {
      if (this.$cookies.get("darkMode")  === "true" ||
          (this.$cookies.get("darkMode") ===  null  &&
           window.matchMedia("(prefers-color-scheme: dark)").matches)) {
        this.$store.commit(TOGGLE_DARK_MODE);
      }
    },
    registerColorSchemeListener() {
      window.matchMedia("(prefers-color-scheme: dark)").addEventListener(
        'change', () => {
          if (this.$cookies.get("darkMode") === null) {
            this.$store.commit(TOGGLE_DARK_MODE);
        }
      });
    }
  },
```

Any ideas?

Reply:
Reads much better! Lets take it one step further.

> **Describe a possible solution**
If there is no darkMode cookie, query the prefers-color-scheme to check the current color scheme, and add the cookie according to the query result.

Lets only read the preferance once with `.matches` and not keep the listener open. This change would sit more inline with the UX that we are looking for from the issue definition

Codewise, we'd no longer need the eventListener as we'd only take the value once that you can wrap in a function `getDarkModeDevicePreferance`.

Then, I think we can follow the earlier logic that reads at the same abstraction level as the issue statement
```js
sync created() {
  this.syncDarkMode();
  ...
}
syncDarkMode() {
  if (this.isDarkModeCookieSet()) {
     this.syncDarkModeWithCookie();
  } else {
     this.syncDarkModeWithDevicePreference(); // get pref, set dark mode, and save cookie
  }
}
```

Also small note on this predicates like this (not sure if itll be the same one):
```js
(this.$cookies.get("darkMode")  === "true" || (this.isFollowingDeviceColor() && mq.matches)
```
but better to give it a function so that when reading you can stay at one level of abstraction (figuring out where darkmode preferances are coming from) and not worry about mq.matches and what that means and what a JS MediaQuery object is.

Lastly, strings `"darkMode:` and the `"(prefers-color-scheme: dark)"` can be extracted out like
https://github.com/YACS-RCOS/yacs.n/blob/master/src/web/src/controllers/Schedule.js#L7-L8
so that we catch string mismatches as compile time errors. You can throw these strings up to the top of the class as something like `App.COOKIE_DARKMODE` and `App.MQ_COLORSCHEME` or `App.MEDIAQUERY_COLORSCHEME`

wyt? or should we keep the responsiveness to colorscheme changes?

Reply:
I'm way out of the loop here, but instead of only using `prefers-color-scheme: dark` when the `darkMode` cookie is null, could we just update the `darkMode` cookie state to match the `prefers-color-scheme: dark` state if it changes?

Reply:
> Lets only read the preferance once with .matches and not keep the listener open. This change would sit more inline with the UX that we are looking for from the issue definition
> 
> Codewise, we'd no longer need the eventListener as we'd only take the value once that you can wrap in a function getDarkModeDevicePreferance.
> 
> Then, I think we can follow the earlier logic that reads at the same abstraction level as the issue statement [...]

Sorry, I forgot to update the feature request.  My bad.  That was the original idea, and it only covers the change for new users who don't have cookies.  On actual coding I found it is easy to implement a responsive color scheme as a third state...  Personally I would like to preserve this feature.  

> Also small note on this predicates like this (not sure if itll be the same one):
> 
> ```javascript
> (this.$cookies.get("darkMode")  === "true" || (this.isFollowingDeviceColor() && mq.matches)
> ```
> 
> but better to give it a function so that when reading you can stay at one level of abstraction (figuring out where darkmode preferances are coming from) and not worry about mq.matches and what that means and what a JS MediaQuery object is.

Agree.  I thought this sentence shows a clear relationship of the conditions...  But not optimal for reading it.  
I will wrap it before the next push.

> Lastly, strings `"darkMode:` and the `"(prefers-color-scheme: dark)"` can be extracted out like [/src/web/src/controllers/Schedule.js: L7-L8](https://github.com/YACS-RCOS/yacs.n/blob/master/src/web/src/controllers/Schedule.js#L7-L8) so that we catch string mismatches as compile time errors. You can throw these strings up to the top of the class as something like `App.COOKIE_DARKMODE` and `App.MQ_COLORSCHEME` or `App.MEDIAQUERY_COLORSCHEME`

Do you mean `state.darkMode` in `store/index.js`?  Not sure if this will affect current usage of cookies... 
Also, not sure if `App.Vue` has a `class`...

Reply:
> I'm way out of the loop here, but instead of only using `prefers-color-scheme: dark` when the `darkMode` cookie is null, could we just update the `darkMode` cookie state to match the `prefers-color-scheme: dark` state if it changes?

It is a bit confusing.  
If I interpret your idea correct, `darkMode` cookie has a string for saving the current user's option for color scheme.  

Also there is a boolean `state.darkMode`, which will record the current site color scheme.  This will change on each load, if the cookie was set to dark mode; or the user did not specify the scheme, it will change to follow the current device preference.

Reply:
Yep you're right, extracting out into `store/index.js` should do it.

Shouldn't change the behavior, only a code change

Proposed change goes from:
```js
return this.$cookies.get("darkMode") === null;
```
to
```js
return this.$cookies.get(App.COOKIE_DARKMODE) === null;
```
so easier to take advantage of autocomplete and reduce typos/silent errors

Reply:
> extracting out into `store/index.js` should do it where you'd import `COOKIE_DARKMODE`.
> Shouldn't change the behavior, only a code change
> 
> Proposed change goes from:
> ```javascript
> return this.$cookies.get("darkMode") === null;
> ```
> to
> ```javascript
> return this.$cookies.get(COOKIE_DARKMODE) === null;
> ```
> so easier to take advantage of autocomplete and reduce typos/silent errors

Perhaps something like this code below?
`store/index.js`: `export const COOKIE_DARK_MODE = "darkMode";`
Same `import` will apply to `Header.Vue` too.

<details>
  <summary><code>App.Vue</code></summary>

```javascript
import { TOGGLE_DARK_MODE, SET_COURSE_LIST, COOKIE_DARK_MODE } from "@/store";
export const DARK_MODE_MEDIA_QUERY = "(prefers-color-scheme: dark)"; 
export default {
  methods: {
    syncColorScheme() {
      if (this.isCookieOrDeviceInDarkMode()) {
        this.$store.commit(TOGGLE_DARK_MODE);
      }
    },
    registerColorSchemeListener() {
      window.matchMedia(DARK_MODE_MEDIA_QUERY).addEventListener(
        "change", () => {
          if (this.isFollowingDeviceColor()) {
            this.$store.commit(TOGGLE_DARK_MODE);
        }
      });
    },
    isFollowingDeviceColor() {
      return this.$cookies.get(COOKIE_DARK_MODE) === null;
    },
    isCookieOrDeviceInDarkMode() {
      return this.$cookies.get(COOKIE_DARK_MODE) === "true" ||
              (this.isFollowingDeviceColor() && 
               window.matchMedia(DARK_MODE_MEDIA_QUERY).matches);
    }
  },
}
```

</details>